File: frontend/lib/sales.ts
import { type Paged, api, encodeId } from "@/lib/http";

export type InvoiceLine = {
  item_code: string;
  /** The item's name as billed. Null on list rows — an ERPNext list query
   *  returns no child table at all — so always fall back to the code. */
  item_name: string | null;
  /** The product's own selling price when the invoice was raised. `rate` is
   *  what was actually charged; the gap between them is the concession. Null
   *  when the product had no price on file. */
  list_rate: number | null;
  qty: number;
  rate: number;
  amount: number;
};

export type SalesInvoice = {
  id: string;
  customer: string;
  posting_date: string | null;
  due_date: string | null;
  grand_total: number;
  outstanding_amount: number;
  status: string;
  remarks: string | null;
  items: InvoiceLine[];
  update_stock: boolean;
  taxes_and_charges: string | null;
  is_commissioned: boolean;
  commission_agent: string | null;
  amended_from: string | null;
  is_cancelled: boolean;
  is_opening: boolean;
};

export type SalesInvoiceInput = {
  customer: string;
  items: { item_code: string; qty: number; rate: number; description?: string | null }[];
  posting_date?: string | null;
  due_date?: string | null;
  remarks?: string | null;
  update_stock?: boolean;
  taxes_and_charges?: string | null;
  is_commissioned?: boolean;
  commission_agent?: string | null;
};

export const listSales = (
  token: string,
  params: Paged & { search?: string; status?: string } = {}
) => api.get<SalesInvoice[]>(token, "/sales", { limit: 200, ...params });

export const getSale = (token: string, id: string) =>
  api.get<SalesInvoice>(token, `/sales/${encodeId(id)}`);

export const createSale = (token: string, input: SalesInvoiceInput) =>
  api.post<SalesInvoice>(token, "/sales", input);

/**
 * Correct a posted invoice.
 *
 * ERPNext cannot edit a submitted document, so the API cancels this one and
 * posts the correction in its place. The invoice **number changes** — the
 * response carries the new id, and the id passed in now names a cancelled
 * document. Callers must read the id back rather than reuse the one they sent.
 *
 * The whole document is replaced, so `input` has to carry every field the
 * invoice should keep, not just the changed ones.
 */
export const updateSale = (token: string, id: string, input: SalesInvoiceInput) =>
  api.put<SalesInvoice>(token, `/sales/${encodeId(id)}`, input);

/**
 * Why this invoice cannot be corrected, or null if it can.
 *
 * Mirrors `SalesService._ensure_amendable`. Nothing here is a control — the API
 * refuses independently — it only decides whether offering the action would end
 * in a refusal.
 */
/** What a commissioned sale gave away: the list price less what was charged,
 *  across the lines that were discounted. Lines with no list price on file
 *  contribute nothing rather than counting as a total giveaway. */
export function concession(invoice: SalesInvoice): number {
  return invoice.items.reduce(
    (sum, l) => sum + (l.list_rate && l.list_rate > l.rate
      ? (l.list_rate - l.rate) * l.qty
      : 0),
    0
  );
}

export function amendBlockedReason(
  invoice: SalesInvoice
): "opening" | "settled" | null {
  if (invoice.is_opening) return "opening";
  // A cancelled invoice reads as fully settled because ERPNext zeroes the
  // outstanding amount when it reverses one; that is not a payment.
  if (invoice.is_cancelled) return null;
  if (invoice.grand_total - invoice.outstanding_amount > 0.005) return "settled";
  return null;
}

File: backend/schemas/sales.py
"""Sales invoice schemas (Pydantic V2). Proxy over the ERPNext Sales Invoice DocType."""

from pydantic import BaseModel, Field


class InvoiceLineInput(BaseModel):
    item_code: str = Field(min_length=1, max_length=140)
    qty: float = Field(gt=0)
    rate: float = Field(ge=0)
    description: str | None = None


class SalesInvoiceCreate(BaseModel):
    customer: str = Field(min_length=1)
    items: list[InvoiceLineInput] = Field(min_length=1)
    due_date: str | None = None
    posting_date: str | None = None
    remarks: str | None = None
    update_stock: bool = False
    taxes_and_charges: str | None = None
    # A sale brokered by an agent and billed below the product's selling price.
    # The customer pays the lower figure and no payout is owed — the agent keeps
    # the spread — so this is a discount with a reason recorded against it, not
    # ERPNext's `sales_partner` commission. Without the reason, a deliberate
    # concession and a mistyped rate look identical a month later.
    is_commissioned: bool = False
    commission_agent: str | None = Field(default=None, max_length=140)


class SalesInvoiceUpdate(SalesInvoiceCreate):
    """Replaces a *posted* invoice wholesale.

    A submitted Sales Invoice is immutable in ERPNext, so this is applied as
    cancel-then-amend: every field is rewritten from this payload exactly as a
    create would write it, and an omitted field is cleared rather than kept.
    Same shape as SalesInvoiceCreate, named separately so the API contract says
    which one a route takes.
    """


class InvoiceLine(BaseModel):
    item_code: str
    # What the product's own selling price was when the invoice was raised
    # (ERPNext `price_list_rate`), against which `rate` is the amount actually
    # charged. The pair is what makes "how much did we give away" answerable;
    # None when the product had no price on file.
    list_rate: float | None = None
    # ERPNext copies the item's name onto the line when the invoice is saved, so
    # this is what the document was actually billed as — not today's catalogue
    # name. Optional because a list query never returns child rows at all.
    item_name: str | None = None
    qty: float
    rate: float
    amount: float


class SalesInvoiceRead(BaseModel):
    id: str
    customer: str
    posting_date: str | None = None
    due_date: str | None = None
    grand_total: float
    outstanding_amount: float
    status: str
    remarks: str | None = None
    items: list[InvoiceLine] = []
    # Round-tripped so an amendment can be posted without losing them. A PUT
    # replaces the whole document, so anything the write model accepts but the
    # read model hides is silently cleared by an edit — dropping the VAT
    # template off an invoice, or leaving stock issued against a cancelled one.
    update_stock: bool = False
    taxes_and_charges: str | None = None
    is_commissioned: bool = False
    commission_agent: str | None = None
    # Amendment lineage. `amended_from` names the cancelled invoice this one
    # replaces; `is_cancelled` marks an invoice that has been reversed and no
    # longer affects the ledger. Both are needed to decide whether an invoice
    # may still be edited, and by what — see SalesService.update.
    amended_from: str | None = None
    is_cancelled: bool = False
    is_opening: bool = False
